components/debounce/browser/debounce_rule.cc

```diff
@@ -86,16 +166,65 @@ bool DebounceRule::Apply(const GURL& original_url, GURL* final_url) const {
   if (!include_pattern_set_.MatchesURL(original_url))
     return false;
 
-  std::string unescaped_value;
-  if (!net::GetValueForKeyInQuery(original_url, param_, &unescaped_value))
-    return false;
-  if ((action_ == kDebounceBase64DecodeAndRedirectToParam) &&
-      (!base::Base64UrlDecode(unescaped_value,
-                              base::Base64UrlDecodePolicy::IGNORE_PADDING,
-                              &unescaped_value))) {
+  if (!DebounceRule::CheckPrefForRule(prefs)) {
     return false;
   }
+
+  std::string unescaped_value;
+
+  if (action_ == kDebounceRegexPath) {
+    // Important: Apply param regex to ONLY the path of original URL.
+    auto path = original_url.path();
+    const re2::RE2 pattern_regex(param_);
+    if (!pattern_regex.ok()) {
+      LOG(WARNING) << "Debounce rule has param: " << param_
+                   << " which is an invalid regex pattern";
+      return false;
+    }
+    if (pattern_regex.NumberOfCapturingGroups() != 1) {
+      LOG(WARNING) << "Debounce rule has param: " << param_
+                   << " which captures != 1 groups";
+      return false;
+    }
+    if (!RE2::PartialMatch(path, pattern_regex, &unescaped_value)) {
+      LOG(INFO) << "Debounce rule with param: " << param_
+                << " was unable to capture string despite match";
+      return false;
+    }
+
+    // unescaped_value now has a string; we will check if the captured value is
+    // a valid URL down below
+  } else {
+    if (!net::GetValueForKeyInQuery(original_url, param_, &unescaped_value))
+      return false;
+    if ((action_ == kDebounceBase64DecodeAndRedirectToParam) &&
+        (!base::Base64UrlDecode(unescaped_value,
+                                base::Base64UrlDecodePolicy::IGNORE_PADDING,
+                                &unescaped_value))) {
+      return false;
+    }
+  }
+
   GURL new_url(unescaped_value);
+  // Important: If there is a prepend_scheme in the rule BUT the URL is already
+  // valid i.e. has a scheme, we treat this as an erroneous rule and do not
+  // apply it.
+  if (prepend_scheme_ && new_url.is_valid()) {
+    LOG(WARNING) << "Debounce rule with param: " << param_
+                 << " and prepend scheme " << prepend_scheme_
+                 << " got a valid URL, treating as erroneous rule";
+    return false;
+  }
+  // If there is a prepend_scheme specified AND the URL is not valid, prepend
+  // the specified scheme and try again
+  if (prepend_scheme_ && !new_url.is_valid()) {
+    std::string scheme = (prepend_scheme_ == kDebounceSchemePrependHttp)
+                             ? url::kHttpScheme
+                             : url::kHttpsScheme;
+    auto new_url_spec =
+        base::StringPrintf("%s://%s", scheme.c_str(), unescaped_value.c_str());
+    new_url = GURL(new_url_spec);
+  }
```

Comment:
Just to be extra sure that the URL was created properly, I'd suggest a `DCHECK()` to check that `new_url.scheme() == scheme`.